### PR TITLE
Elementizes plastic flap hitbox behavior into the Render over, Keep hitbox element! Applies to directional windows

### LINKED ISSUE #110: Directional windows layer under humans when it doesn't make sense, should probably frill as well?
I am on the same tile as a south facing directional window, yet I show as being over it.

image

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request

Plastic flaps render over people, but you're still able to hit the person crawling under. This is exactly what directional windows need to act properly, so I've elementized the behavior and applied to both!

NOTE: Plastic flap code does pic related, and so this element brings that visual error with it. But it's still worth!
image

## Why It's Good For The Game

Fixes #110 

## Changelog

:cl:
fix: Directional windows now properly render over you
/:cl:

### CODE REVIEW

File: code/datums/elements/render_over_keep_hitbox.dm [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+	RegisterSignal(obj_target, COMSIG_MOVABLE_Z_CHANGED, .proc/on_changed_z_level)
+	obj_target.alpha = 0
+	gen_overlay(obj_target)
+
+/datum/element/render_over_keep_hitbox/Detach(obj/structure/target, ...)
+	UnregisterSignal(target, COMSIG_MOVABLE_Z_CHANGED)
+	target.alpha = initial(target.alpha)
+	SSvis_overlays.remove_vis_overlay(target.managed_vis_overlays)
+	. = ..()
```
Comment (MrMelbert) [inline code/datums/elements/render_over_keep_hitbox.dm]:
```suggestion
	return ..()
```

File: code/datums/elements/render_over_keep_hitbox.dm [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * # Render over, Keep hitbox element!
+ *
+ * Non bespoke element (1 in existence) that makes structures render over mobs, but still allow you to attack the mob's hitbox!
+ * Used in plastic flaps, and directional windows!
+ */
+/datum/element/render_over_keep_hitbox
+
+/datum/element/render_over_keep_hitbox/Attach(datum/target)
+	. = ..()
+	if(!istype(target, /obj/structure))
```
Comment (MrMelbert) [inline code/datums/elements/render_over_keep_hitbox.dm]:
```suggestion
	if(!isstructure(target))
```

File: code/datums/elements/render_over_keep_hitbox.dm [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	gen_overlay(obj_target)
+
+/datum/element/render_over_keep_hitbox/Detach(obj/structure/target, ...)
+	UnregisterSignal(target, COMSIG_MOVABLE_Z_CHANGED)
+	target.alpha = initial(target.alpha)
+	SSvis_overlays.remove_vis_overlay(target.managed_vis_overlays)
+	. = ..()
+
+/datum/element/render_over_keep_hitbox/proc/on_changed_z_level(obj/structure/target, turf/old_turf, turf/new_turf, same_z_layer)
+	if(same_z_layer)
```
Comment (MrMelbert) [inline code/datums/elements/render_over_keep_hitbox.dm]:
```suggestion
/datum/element/render_over_keep_hitbox/proc/on_changed_z_level(obj/structure/target, turf/old_turf, turf/new_turf, same_z_layer)
	SIGNAL_HANDLER
	
	if(same_z_layer)
```

File: code/datums/elements/render_over_keep_hitbox.dm [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+		return
+	SSvis_overlays.remove_vis_overlay(target.managed_vis_overlays)
+	gen_overlay(target)
+	return
```
Comment (MrMelbert) [inline code/datums/elements/render_over_keep_hitbox.dm]:
```suggestion
```

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/datums/elements/ranged_attacks.dm [hunk 5]
```diff
@@ -9,7 +9,7 @@
 /datum/element/ranged_attacks/Attach(atom/movable/target, casingtype, projectilesound, projectiletype)
 	. = ..()
 	if(!isbasicmob(target))
-		return COMPONENT_INCOMPATIBLE
+		return ELEMENT_INCOMPATIBLE
```
Comment (LemonInTheDark) [inline code/datums/elements/ranged_attacks.dm]: not here, I don't want misc changes out of risk of more merge conflicts

File: code/datums/elements/render_over_keep_hitbox.dm [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	//you see mobs under it, but you hit them like they are above it
+	SSvis_overlays.add_vis_overlay(
+		target,
+		target.icon,
+		target.icon_state,
+		ABOVE_MOB_LAYER,
+		MUTATE_PLANE(GAME_PLANE, our_turf),
+		target.dir,
+		add_appearance_flags = RESET_ALPHA
+	)
```
Comment (LemonInTheDark) [inline code/datums/elements/render_over_keep_hitbox.dm:44]: do we need to use vis_contents for this to work?
~~also how does this work~~

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

**Comment by LemonInTheDark** [comment]
> NOTE: Plastic flap code does pic related, and so this element brings that visual error with it. But it's still worth! ![image]

This is an issue of sidemap.
Basically since everything (the window and the mob) are on the same plane, we decide rendering order based off (oversimplifying)
"position (x + pixel_x, bounded to blocks of 32x32)"
failing that, layer

This means you can in theory construct a window that renders below when we want it to, and above when we want it to, which is what we want here.
Getting it to be click transparent is another issue, not sure how you go about that.

**Comment by tralezab** [comment]
mkay, i'm back

**Pull request merged**